### PR TITLE
Redesign of MSA retirement to use signed extension

### LINKED ISSUE #340: MSA Retirement should be Free
## Related
- https://github.com/LibertyDSNP/frequency/issues/44

Should be able to follow the pattern from https://github.com/LibertyDSNP/frequency/pull/323 

As a user, I want to be able to retire my MSA, even if I have _zero_ tokens.

## Acceptance Criteria
- [x] Given that an MSA has **one** key, I should be able to retire my MSA without any tokens.

## Notes
- #44
- #205
- #229 
- https://github.com/LibertyDSNP/frequency/pull/323

### PR DESCRIPTION
**Pull request by rlaferla**
Moved checks to CheckFreeExtrinsicUse signed extension for MSA retirement.

Closes #340

### CODE REVIEW

File: pallets/msa/src/tests.rs [hunk 1]
```diff
@@ -437,36 +437,6 @@ fn test_retire_msa_success() {
 	})
 }
 
-#[test]
```
Comment (shannonwells) [inline pallets/msa/src/tests.rs:440]: We have tests for the MSA pallet signed extensions already, these can simply be converted to a signed extension test and unit tests for `validate_msa_retirement` and would be better than deleting them.

**Review by shannonwells — CHANGES_REQUESTED** [review]
Other than keeping the tests around, looks good.

File: pallets/msa/src/tests.rs (hunk 1 above)
Comment (rlaferla) [reply pallets/msa/src/tests.rs:440]: Agreed. I will update.

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -1246,14 +1237,42 @@ impl<T: Config + Send + Sync> CheckFreeExtrinsicUse<T> {
 			.into();
 		return ValidTransaction::with_tag_prefix(TAG_PREFIX).and_provides(account_id).build()
 	}
+
+	/// validates that a MSA being retired is valid
+	pub fn validate_msa_retirement(account_id: &T::AccountId) -> TransactionValidity {
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:1242]: to continue the pattern of using ensure, perhaps, call it 
```
ensure_msa_can_retire
```

File: pallets/msa/src/lib.rs [hunk 3]
```diff
@@ -1246,14 +1237,42 @@ impl<T: Config + Send + Sync> CheckFreeExtrinsicUse<T> {
...
+			)
+		);
+
+		// Invalid transaction error "MoreThanOneKeyExists" if the MSA has more than one account key.
+		let key_count = Pallet::<T>::get_public_key_count_by_msa_id(msa_id);
+		ensure!(
+			key_count == 1,
+			InvalidTransaction::Custom(ValidityError::InvalidMoreThanOneKeyExists as u8)
+		);
+		return ValidTransaction::with_tag_prefix(TAG_PREFIX).and_provides(account_id).build()
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:1262]: I don't think you need the return statement.

File: pallets/msa/src/lib.rs [hunk 4]
```diff
@@ -1246,14 +1237,42 @@ impl<T: Config + Send + Sync> CheckFreeExtrinsicUse<T> {
...
 	}
+
+	/// validates that a MSA being retired is valid
+	pub fn validate_msa_retirement(account_id: &T::AccountId) -> TransactionValidity {
+		const TAG_PREFIX: &str = "MSARetirement";
+		let msa_id = Pallet::<T>::ensure_valid_msa_key(account_id)
+			.map_err(|_| InvalidTransaction::Custom(ValidityError::InvalidMsaKey as u8))?
+			.into();
+
+		// Invalid transaction error "InvalidRegisteredProviderCannotBeRetired" if the MSA id is a registered provider
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:1248]: can you please explain the value that adding a comment here provides that is not easily deduced by the line below?

File: pallets/msa/src/lib.rs (hunk 2 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:1242]: Sure

File: pallets/msa/src/lib.rs (hunk 3 above)
Comment (rlaferla) [reply pallets/msa/src/lib.rs:1262]: Removed return.

**Comment by rlaferla** [comment]
Using new branch (includes resolved comments.)

**Pull request closed without merging**